jconstraints-core/src/main/java/gov/nasa/jpf/constraints/solvers/encapsulation/ProcessWrapperSolver.java

```diff
@@ -79,28 +90,50 @@ public ProcessWrapperSolver(String solver, String javaBinary) {
     }
   }
 
-  @Override
-  public synchronized Result solve(Expression<Boolean> f, Valuation result) {
+  Result solve(List<Expression<Boolean>> f, Valuation result, int counter) {
+
     try {
       return runSolverProcess(f, result);
     } catch (IOException | ClassNotFoundException e) {
       logCallToSolver(f);
       e.printStackTrace();
+      ++counter;
+      if (counter < RETRIES) {
+        if (solver != null && solver.isAlive()) {
+          solver.destroy();
+        }
+        solver = null;
+        bes = null;
+        bos = null;
+        inObject = null;
+        solve(f, result, counter);
+      }
       return Result.DONT_KNOW;
     } catch (InterruptedException e) {
       solver = null;
       outObject = null;
       System.out.println("Restart required");
-      return solve(f, result);
+      return solve(f, result, 0);
     }
   }
 
+  @Override
+  public Result solve(Expression<Boolean> f, Valuation result) {
+    List<Expression<Boolean>> tmp = new LinkedList<>();
+    tmp.add(f);
+    return solve(tmp, result, 0);
+  }
+
   @Override
   public SolverContext createContext() {
-    return new ProcessWrapperContext(solverName, javaBinary);
+    UNSATCoreSolver ctx = new ProcessWrapperContext(solverName, javaBinary);
+    if (isUnsatTracking) {
+      ctx.enableUnsatTracking();
+    }
+    return (ProcessWrapperContext) ctx;
```

Comment:
Could be simplified by not upcasting
```suggestion
    ProcessWrapperContext ctx = new ProcessWrapperContext(solverName, javaBinary);
    if (isUnsatTracking) {
      ctx.enableUnsatTracking();
    }
    return ctx;
```